### ISSUE TITLE
Sending cancel safety

### ISSUE THREAD

**Issue by BigBigos** [smol-rs/async-channel#87]
I believe `Sender::send()` is not cancel safe in regard that it takes ownership of the value to be sent. If the future is cancelled, the value is gone with no way to recover it. This makes using it inconvenient within the `select!` loop.

tokio's `mpsc::Sender` suggests using `reserve()` and `Permit::send()` to resolve this problem. `reserve()` returns a future, but it does not consume any value by itself. `Permit::send()` then can be used to send the value without any blocking.

Can a similar API be implemented for async-channel? Is that feasible?

[1] https://docs.rs/tokio/latest/tokio/macro.select.html
[2] https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#cancel-safety

**Comment by notgull** [comment]
Support would need to be added to the underlying `concurrent-queue` crate... which may be a lot harder in practice. I'd accept PRs but we would first need to come up with the underlying algorithm.

**Comment by samuelhnrq** [comment]
Sorry for necro bumping, but it is the same question, has the situation had any change? Is the library cancel safe as of now? More specifically Im trying to select! an recv vs a sleep (basically a timeout) is there a chance of a race condition of messages being lost (e.g. time running out exactly as the message gets recv'ed and lost)

**Comment by notgull** [comment]
No, we're not truly cancel safe yet. As mentioned this needs to be fixed in `concurrent-queue` first.
